### ISSUE TITLE
HTML Web Resource DOM not consistently loaded when New/Disregard changes is used.

### ISSUE THREAD

**Issue by brandonwoz** [MicrosoftDocs/powerapps-docs#5239]
Issue Overview:
In a Model Driven App, we have a JavaScript OnLoad event handler that calls a method on a content window, which is responsible for modifying the DOM of an HTML web resource. The content window method takes parameters to dictate the DOM updates, specifically a button id, visibility property, and call back function. This setup works as expected on the initial page load. (The DOM is loaded, and the button is updated with the correct attributes and event listener). However, issues arise when the user interacts with the "New" button on the ribbon and then disregards changes to the form.

Specific Problem:
Occasionally, after the user disregards changes, the DOM reloads after the content window method invocation, erasing the client-side state set by the OnLoad event handler. This behavior is intermittent and unpredictable, which complicates debugging and user experience.

Attempted Solution:
To address this, I implemented a content window method that returns a promise, intended to resolve once the DOM is fully loaded. The idea was for the OnLoad event handler to wait for this promise before proceeding with the data transmission for client-side updates. However, this approach has proven inconsistent. There are instances where the DOM is deemed "completed," the promise resolves, and the data is sent, but then the DOM reloads, leading to the same issue of the wiped client-side state.

Request for Assistance:
I am seeking insights or solutions to ensure the client-side state is reliably maintained, especially after user interactions that might trigger a DOM reload. Any advice on handling such scenarios, possibly with better event handling or state management techniques, would be greatly appreciated.

Steps to Reproduce:
Load the Model Driven App page.
Interact with the OnLoad event handler (e.g., through the "New" button on the ribbon).
Disregard changes to the form.
Observe the intermittent DOM reload and the loss of client-side state (causing the button to not update)

Example Code

OnLoad Event Handler 
```

export async function handleCustomButtonTest( executionContext: Xrm.Events.EventContext) {
    const formContext = executionContext.getFormContext();
    const controlName = "WebResource_custom_button_test";
    const control = formContext.getControl(controlName) as Xrm.Controls.IframeControl;
    control.setVisible(true);
    const contentWindow = (await control.getContentWindow()) as IContentWindow;
    if (contentWindow) {
        const formType = formContext.ui.getFormType();
        if (formType === 1) {
            const buttonId = 'saveButton';
            const isVisible = true;
            contentWindow.domIsLoaded().then(() => {
                contentWindow.setCallbackAndVisibility(
                    buttonId,
                    isVisible,
                    saveRequest
                );
            });
        } else {
            // do something else
        }
    }
}
```

Client-side/Content window code
```
buttonLabels = {
    saveButton: "Save"
};
function domIsLoaded() {
    return new Promise((resolve) => {
        if (document.readyState === "loading") {
            document.addEventListener("DOMContentLoaded", resolve);
        } else {
            resolve();
        }
    });
}
function setCallbackAndVisibility(btnId, isVisible, cb) {
    window.buttonId = btnId;
    window.buttonIsVisible = isVisible;
    window.buttonCb = cb;
    updateButtons();
}
function updateButtons() {
    const button = document.getElementById(window.buttonId);
    button.style.display = window.buttonIsVisible ? "block" : "none":;
    button.innerText = buttonLabels[window.buttonId];
    button.addEventListener('click', window.buttonCb);
}
```

**Comment by blakeZTL** [comment]
Yes I have had the same issues and havent had a chance to dive in like you. Thank you so much and I hope this issue can be addressed.

**Comment by JimDaly** [comment]
@brandonwoz  and @blakeZTL 

This repository is for documentation and issues opened for it should be about the documentation. 

What part of the documentation needs to be changed to address your concern? Are any of [our APIs](https://learn.microsoft.com/power-apps/developer/model-driven-apps/clientapi/reference) not working as documented? Please be specific and link to the specific reference page for that API.

Please help me transform your feedback into some specific action we can take to improve the documentation.

I can accept the feedback that you are looking for guidance _on handling such scenarios, possibly with better event handling or state management techniques_.

Have you found the content here: [Troubleshoot form issues in model-driven apps](https://learn.microsoft.com/power-apps/developer/model-driven-apps/troubleshoot-forms)? Does it help?

Have you tried to [Use Monitor to troubleshoot model-driven app form behavior](https://learn.microsoft.com/en-us/power-apps/maker/model-driven-apps/monitor-form-checker)

Otherwise, it looks like you are looking for assistance resolving a technical support issue. We don't have the resources to troubleshoot this with you. I recommend you contact technical support with your repro steps.

**Comment by jarvan-jiang** [comment]
Not sure if it is a right place to discuss the issue, but that is an annoying issue and I can't find any useful discussion in other places.
We also faced similar issues of this kind of unexpected reload, but in our case it occurred after save.
The mitigate solution is that, we observed this kind of reload often carried incomplete parameters, thus we prevent this kind of event to run onload again, by add this code in the beginning of onload:
```
var formContext = executionContext.getFormContext();            
if (!formContext.ui.getFormDescriptor()) { return; }
``` 
Hope this helps.
